### PR TITLE
Closes #50 Added special test case supported Windows

### LINKED ISSUE #50: Long file names, not supported characters on windows
image pasted at 2018-9-4 13-02
image pasted at 2018-9-4 14-12

I have a problems with file names. Request with fix is created :)

### PR DESCRIPTION
**Pull request by AndreyMashukov**
Fixed #50, Created `Spatie\Snapshots\SnapshotTestCase`

### CODE REVIEW

File: src/SnapshotTestCase.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Class SnapshotTestCase.
+ *
+ * Special test case, supported Windows OS.
+ *
+ * @see {https://docs.microsoft.com/en-us/windows/desktop/fileio/naming-a-file}
+ */
+abstract class SnapshotTestCase extends TestCase
+{
+    const WINDOWS_FILE_NAME_LENGTH_LIMIT = 255;
```
Comment (sebastiandedeyne) [inline src/SnapshotTestCase.php]: No need to tie this functionality to Windows, please rename this to `FILE_NAME_LENGTH_LIMIT`. This should be configurable.

Change this property to `protected static $fileNameLengthLimit`, and add a `public static fileNameLengthLimit(?int $limit)` function. By default, the limit should be `null`, and the check should be skipped.

File: src/SnapshotTestCase.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * @throws \ReflectionException
+     * @throws \Exception
+     *
+     * @return string
+     */
+    protected function getSnapshotId(): string
+    {
+        $name = (new \ReflectionClass($this))->getShortName().'_'.
```
Comment (sebastiandedeyne) [inline src/SnapshotTestCase.php]: Import this class above instead of prefixing it with a trailing slash.

File: src/SnapshotTestCase.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            $this->getName().'_'.
+            $this->snapshotIncrementor;
+
+        $replaced = \preg_replace('/(\s+|\W+)/ui', '_', $name);
```
Comment (sebastiandedeyne) [inline src/SnapshotTestCase.php]: No need for trailing slashed before function names.

File: src/SnapshotTestCase.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        if (\mb_strlen($replaced) >= self::WINDOWS_FILE_NAME_LENGTH_LIMIT) {
+            throw new \Exception('Very long filename, please rename test or test case name!');
```
Comment (sebastiandedeyne) [inline src/SnapshotTestCase.php]: Use a custom exception class

File: composer.json [hunk 5]
```diff
@@ -21,7 +21,8 @@
     ],
     "require": {
         "php": "^7.0",
-        "phpunit/phpunit": "^6.0.5|^7.0"
+        "phpunit/phpunit": "^6.0.5|^7.0",
+        "ext-mbstring": "*"
```
Comment (sebastiandedeyne) [inline composer.json]: Is this necessary? Do we really need to rely on an extension for this? What's wrong with `strlen`?

File: src/SnapshotTestCase.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class SnapshotTestCase.
+ *
+ * Special test case, supported Windows OS.
+ *
+ * @see {https://docs.microsoft.com/en-us/windows/desktop/fileio/naming-a-file}
+ */
+abstract class SnapshotTestCase extends TestCase
```
Comment (sebastiandedeyne) [inline src/SnapshotTestCase.php]: There's no need to introduce this abstract class. Move everything to the existing trait.

**Comment by AndreyMashukov** [comment]
Напиши сам если столько правок!

**Comment by sebastiandedeyne** [comment]
> Напиши сам если столько правок!

We're not going to roll over and accept PR's of code we don't want to maintain, especially for a feature we don't need ourselves. 

We're willing to maintain this functionality, if implemented in a generic way and follows the overall code style of the rest of the repository. This PR had red flags on both fronts.

Lastly, please communicate in english when responding to issues or PR's in our repositories.

**Pull request closed without merging**